pkgs/applications/science/math/R/default.nix

```diff
@@ -4,6 +4,7 @@
 , curl, Cocoa, Foundation, libobjc, libcxx, tzdata
 , withRecommendedPackages ? true
 , enableStrictBarrier ? false
+, javaSupport ? (!stdenv.hostPlatform.isAarch32 && !stdenv.hostPlatform.isAarch64)
```

Comment:
This is just a license preference leaked to the code
Here should be something like ```javaSupport ? jdk.meta.available```, in order to cover all architectures and to enable ```javaSupport``` when ```config.allowUnfree == true```

Reply:
I am okay with either way here, but some have objected in the past to using meta.available (#45727). I think there are also some cases where meta.available breaks for some unclear reason. I think the "correct" way to do this would be to use meta.platforms:

```
stdenv.lib.any (stdenv.lib.meta.platformMatch stdenv.hostPlatform) jdk.meta.platforms
```

But as I said I think either way is okay.

Reply:
Nixpkgs doesn't really seem to support optional dependencies. Having a javaSupport parameter works fine, but doesn't seem to be the norm. Indeed, having a similar parameter for every optional dependency would add a lot of complexity. That said, removing the repeated conditionals and replacing them with a named parameter is easier to understand and tells you why something is conditional.

I assume that once a version of OpenJDK is packaged for aarch32/64, we'll revert the R package to depend on jdk once more.

Reply:
```R``` with the only ```jdk``` dependency is rather a simple case.
But there are many others.
For example, take a look at [collectd](https://github.com/NixOS/nixpkgs/blob/57bccb3cb8f6b5dea5c315fcaabdb992498b9314/pkgs/tools/system/collectd/default.nix).
Besides optional ```jdk``` dependency (with the same problem as the one here with ```R```) it has dozens of other dependencies, and almost all of them are optional.
On every combination of ```(system, config.allowUnfree, config.allowBroken, ...)```, some of them are not available making ```collectd``` derivation broken too.

So:
 * Nobody would support those lists manually, there are too many of them and some of them are bigger than one ```jdk```
 * ```meta.available``` declared "bad" because might trigger rebuild for non-standard configuration
 * The only practical way to use non-broken ```collectd``` is
```
  services.collectd.package = pkgs.collectd.override {
    jdk = if (!stdenv.hostPlatform.isAarch32 && !stdenv.hostPlatform.isAarch64) pkgs.jdk else null;
   # ~20 other ifs
  };
```
This is actually what my configuration looks like; I am not sure if it is better than ```meta.available```, it also requires rebuilds tailored for each platform (sometimes each machine)

Reply:
I didn't actually try to build R using override to explicitly set jdk to null. Would it have worked? I.e. are null buildInputs ignored? and what does "JAVA_HOME=${jdk}"  do if jdk is null?

Reply:
> I.e. are null buildInputs ignored? 

yes, and many derivations depend on it (I find it a weirder hack than ```meta.available```).

> and what does "JAVA_HOME=${jdk}" do if jdk is null?

Probaly a runtime error, other derivations use something like ```optionalString (jdk != null) "JAVA_HOME=${jdk}"```

Reply:
> optionalString (jdk != null) "JAVA_HOME=${jdk}"

So having to replace "JAVA_HOME=${jdk}" with something like this effectively stops you using override to drop optional dependencies unless it's been explicitly handled in the derivation.